### PR TITLE
Optimize parallelize_loop by using the main thread too

### PR DESCRIPTION
**Pull request by stweil**
Signed-off-by: Stefan Weil <sw@weilnetz.de>

### CODE REVIEW

File: thread_pool.hpp [hunk 1]
```diff
@@ -144,18 +144,22 @@ class thread_pool
...
             num_blocks = (ui32)total_size > 1 ? (ui32)total_size : 1;
         }
+        num_blocks--;
         std::atomic<ui32> blocks_running = 0;
         for (ui32 t = 0; t < num_blocks; t++)
         {
             T start = ((T)(t * block_size) + the_first_index);
-            T end = (t == num_blocks - 1) ? last_index + 1 : ((T)((t + 1) * block_size) + the_first_index);
+            T end = ((T)((t + 1) * block_size) + the_first_index);
             blocks_running++;
```
Comment (stweil) [inline thread_pool.hpp:153]: This line would not be needed if `blocks_running` were initialized to `num_blocks`.

**Comment by bshoshany** [comment]
Thanks for the suggestion!

But I don't think this is a good idea. Right now the loop tasks are queued in the pool, and only run once all previously queued tasks finish running, which is usually the desired behavior. With the change you are suggesting, one loop task will run immediately, ahead of any tasks previously queued in the pool, while the other loop tasks will be queued in the pool and might be executed only much later.

Furthermore, if the pool has as many threads as the hardware allows, then queued tasks will actually be executed on the main thread as well while it's idling. You can verify this by noticing that when `parallelize_loop()` is running, the task manager will show all cores being used.

**Comment by stweil** [comment]
That's not correct. The main thread and the pool threads will run in parallel right from the beginning. The task manager and the timing results show this when the test application is run.

There is also no difference regarding the hardware usage. If there is a main thread with n pool threads, the old code will run n pool threads and a polling main thread, while the new code will run (n - 1) pool thread and a main thread plus a polling pool thread.

The only difference is that the new code can achieve the same performance with a smaller pool (n - 1), and it does not waste time in a polling thread while all threads are busy with the loop task when this smaller pool is used.

**Comment by bshoshany** [comment]
First of all, the issue still remains that if there are 100 tasks currently in the queue, then with your code, one of the loop tasks will be executed immediately in the main thread, while all the other loop tasks will wait until the 100 previous queued tasks finish running (which could be a long time). This goes against the idea of a thread pool; all tasks must be queued and executed in order, the pool shouldn't execute a new task before the previous tasks have finished executing.

Second, in the new version of the thread pool (which I am working on right now), I plan to make `parallelize_loop()` non-blocking, i.e. you will get a future and the main thread could do other stuff while the loop is being calculated, as with `submit()`. So the change you are proposing does not fit into that model.

That said, if this does improve performance significantly, then I might consider adding it as a separate member function. But I would like to see proof for the improved performance first. Would you be able to attach a complete program comparing the existing `parallelize_loop()` with your modified one and demonstrating the improved performance, so that I can test it on my side? Thanks.

**Pull request closed without merging**